zapcore/write_syncer_test.go

```diff
@@ -56,6 +56,16 @@ func TestAddSyncWriteSyncer(t *testing.T) {
 	assert.Error(t, ws.Sync(), "Expected to propagate errors from concrete type's Sync method.")
 }
 
+func TestLockedWriteSyncerInterface(t *testing.T) {
+	buf := &bytes.Buffer{}
+	ws := AddSync(buf)
+	var lws WriteSyncer = &lockedWriteSyncer{ws: ws}
```

Comment:
It'd be even better if the test uses a concrete type separately implemented from the internal `lockedWriteSyncer` since, as I understand it, the entire point of this PR is to support external implementations (and avoid double-locking them).